File: frontend/src/types/reports.ts
/**
 * Reports Types
 *
 * TypeScript types for report builder and export.
 */

// Enums

export type ReportFormat = 'pdf' | 'html' | 'excel' | 'pptx' | 'image';

export type PageSize = 'a4' | 'letter' | 'legal' | 'a3' | 'tabloid';

export type PageOrientation = 'portrait' | 'landscape';

export type ElementType =
  | 'chart'
  | 'table'
  | 'text'
  | 'image'
  | 'kpi'
  | 'heading'
  | 'divider'
  | 'spacer'
  | 'page_break'
  | 'table_of_contents';

export type TextAlignment = 'left' | 'center' | 'right' | 'justify';

// Configuration Types

export interface BrandingConfig {
  logo_url?: string;
  logo_width?: number;
  logo_position?: 'left' | 'center' | 'right';
  primary_color?: string;
  secondary_color?: string;
  accent_color?: string;
  font_family?: string;
  header_font_family?: string;
  company_name?: string;
  footer_text?: string;
}

export interface PageMargins {
  top: number;
  bottom: number;
  left: number;
  right: number;
}

export interface HeaderFooterConfig {
  show_header?: boolean;
  show_footer?: boolean;
  header_height?: number;
  footer_height?: number;
  include_logo?: boolean;
  include_title?: boolean;
  include_date?: boolean;
  include_page_numbers?: boolean;
  custom_header_html?: string;
  custom_footer_html?: string;
}

export interface PageConfig {
  size?: PageSize;
  orientation?: PageOrientation;
  margins?: PageMargins;
  header_footer?: HeaderFooterConfig;
}

// Element Types

export interface TextStyle {
  font_size?: number;
  font_weight?: 'normal' | 'bold';
  font_style?: 'normal' | 'italic';
  color?: string;
  alignment?: TextAlignment;
  line_height?: number;
}

export interface ChartElement {
  element_type: 'chart';
  chart_id: string;
  title?: string;
  show_title?: boolean;
  width?: string;
  height?: number;
  filters?: Record<string, unknown>;
}

export interface TableElement {
  element_type: 'table';
  data_source: string;
  query?: string;
  columns?: string[];
  title?: string;
  show_title?: boolean;
  max_rows?: number;
  show_row_numbers?: boolean;
  striped?: boolean;
  bordered?: boolean;
  compact?: boolean;
}

export interface TextElement {
  element_type: 'text';
  content: string;
  style?: TextStyle;
  markdown?: boolean;
}

export interface HeadingElement {
  element_type: 'heading';
  text: string;
  level?: number;
  style?: TextStyle;
  include_in_toc?: boolean;
}

export interface ImageElement {
  element_type: 'image';
  url: string;
  alt_text?: string;
  width?: string;
  height?: string;
  alignment?: TextAlignment;
}

export interface KPIElement {
  element_type: 'kpi';
  kpi_id?: string;
  label: string;
  value?: string | number;
  query?: string;
  format?: string;
  comparison_value?: number;
  comparison_label?: string;
  trend_direction?: 'up' | 'down' | 'neutral';
  color?: string;
}

export interface DividerElement {
  element_type: 'divider';
  style?: 'solid' | 'dashed' | 'dotted';
  color?: string;
  thickness?: number;
  margin_top?: number;
  margin_bottom?: number;
}

export interface SpacerElement {
  element_type: 'spacer';
  height?: number;
}

export interface PageBreakElement {
  element_type: 'page_break';
}

export interface TOCElement {
  element_type: 'table_of_contents';
  title?: string;
  max_depth?: number;
  show_page_numbers?: boolean;
}

export type ReportElement =
  | ChartElement
  | TableElement
  | TextElement
  | HeadingElement
  | ImageElement
  | KPIElement
  | DividerElement
  | SpacerElement
  | PageBreakElement
  | TOCElement;

// Report Section

export interface ReportSection {
  id: string;
  title?: string;
  elements: ReportElement[];
  columns?: number;
  background_color?: string;
  padding?: number;
}

// Template Types

export interface ReportTemplate {
  id: string;
  name: string;
  description?: string;
  page_config: PageConfig;
  branding: BrandingConfig;
  sections: ReportSection[];
  workspace_id?: string;
  created_by: string;
  created_at: string;
  updated_at: string;
  is_public: boolean;
  usage_count: number;
  tags: string[];
}

export interface ReportTemplateCreate {
  name: string;
  description?: string;
  page_config?: PageConfig;
  branding?: BrandingConfig;
  sections: ReportSection[];
  workspace_id?: string;
  is_public?: boolean;
  tags?: string[];
}

export interface ReportTemplateUpdate {
  name?: string;
  description?: string;
  page_config?: PageConfig;
  branding?: BrandingConfig;
  sections?: ReportSection[];
  is_public?: boolean;
  tags?: string[];
}

// Report Generation

export interface ReportGenerateRequest {
  template_id?: string;
  dashboard_id?: string;
  title?: string;
  subtitle?: string;
  format?: ReportFormat;
  page_config?: PageConfig;
  branding?: BrandingConfig;
  sections?: ReportSection[];
  filters?: Record<string, unknown>;
  parameters?: Record<string, unknown>;
  include_cover_page?: boolean;
  include_toc?: boolean;
  generated_date?: string;
}

export interface ReportGenerateResponse {
  job_id: string;
  status: 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';
  format: ReportFormat;
  message?: string;
  download_url?: string;
  expires_at?: string;
}

export interface ReportJob {
  id: string;
  template_id?: string;
  dashboard_id?: string;
  title: string;
  format: ReportFormat;
  status: 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';
  progress: number;
  created_by: string;
  created_at: string;
  started_at?: string;
  completed_at?: string;
  error_message?: string;
  download_url?: string;
  file_size?: number;
  expires_at?: string;
}

// Export Types

export interface DashboardExportRequest {
  dashboard_id: string;
  format?: ReportFormat;
  title?: string;
  include_filters?: boolean;
  filter_state?: Record<string, unknown>;
}

export interface ChartExportRequest {
  chart_id: string;
  format?: 'png' | 'jpeg' | 'svg' | 'pdf';
  width?: number;
  height?: number;
  scale?: number;
  include_title?: boolean;
  include_legend?: boolean;
  background_color?: string;
}

// Scheduled Reports

export interface ScheduledReportConfig {
  template_id: string;
  title: string;
  format?: ReportFormat;
  filters?: Record<string, unknown>;
  parameters?: Record<string, unknown>;
}

export interface ScheduledReport {
  id: string;
  name: string;
  config: ScheduledReportConfig;
  schedule: string;
  recipients: string[];
  enabled: boolean;
  workspace_id?: string;
  created_by: string;
  created_at: string;
  last_generated_at?: string;
  next_generation_at?: string;
}

// Template Library

export interface TemplateCategory {
  id: string;
  name: string;
  description?: string;
  icon?: string;
  template_count: number;
}

export interface TemplatePreview {
  id: string;
  name: string;
  description?: string;
  thumbnail_url?: string;
  category?: string;
  tags: string[];
  usage_count: number;
  is_public: boolean;
}

// Format Info

export interface FormatInfo {
  format: ReportFormat;
  name: string;
  description: string;
  extension: string;
  supports_charts: boolean;
  supports_tables: boolean;
  supports_images: boolean;
}

// Constants

export const FORMAT_LABELS: Record<ReportFormat, string> = {
  pdf: 'PDF',
  html: 'HTML',
  excel: 'Excel',
  pptx: 'PowerPoint',
  image: 'Image',
};

export const FORMAT_ICONS: Record<ReportFormat, string> = {
  pdf: 'file-text',
  html: 'globe',
  excel: 'table',
  pptx: 'presentation',
  image: 'image',
};

export const FORMAT_EXTENSIONS: Record<ReportFormat, string> = {
  pdf: '.pdf',
  html: '.html',
  excel: '.xlsx',
  pptx: '.pptx',
  image: '.png',
};

export const PAGE_SIZE_LABELS: Record<PageSize, string> = {
  a4: 'A4',
  letter: 'Letter',
  legal: 'Legal',
  a3: 'A3',
  tabloid: 'Tabloid',
};

export const ORIENTATION_LABELS: Record<PageOrientation, string> = {
  portrait: 'Portrait',
  landscape: 'Landscape',
};

export const ELEMENT_TYPE_LABELS: Record<ElementType, string> = {
  chart: 'Chart',
  table: 'Table',
  text: 'Text',
  image: 'Image',
  kpi: 'KPI Card',
  heading: 'Heading',
  divider: 'Divider',
  spacer: 'Spacer',
  page_break: 'Page Break',
  table_of_contents: 'Table of Contents',
};

export const ELEMENT_TYPE_ICONS: Record<ElementType, string> = {
  chart: 'bar-chart-2',
  table: 'table',
  text: 'type',
  image: 'image',
  kpi: 'activity',
  heading: 'heading',
  divider: 'minus',
  spacer: 'square',
  page_break: 'file-minus',
  table_of_contents: 'list',
};

// Helper Functions

export function getFormatLabel(format: ReportFormat): string {
  return FORMAT_LABELS[format] || format.toUpperCase();
}

export function getFormatIcon(format: ReportFormat): string {
  return FORMAT_ICONS[format] || 'file';
}

export function getFormatExtension(format: ReportFormat): string {
  return FORMAT_EXTENSIONS[format] || '.pdf';
}

export function getElementTypeLabel(type: ElementType): string {
  return ELEMENT_TYPE_LABELS[type] || type;
}

export function getElementTypeIcon(type: ElementType): string {
  return ELEMENT_TYPE_ICONS[type] || 'square';
}

export function formatFileSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

export function getJobStatusColor(status: ReportJob['status']): string {
  const colors: Record<ReportJob['status'], string> = {
    pending: 'gray',
    processing: 'blue',
    completed: 'green',
    failed: 'red',
    cancelled: 'gray',
  };
  return colors[status];
}

export function getJobStatusLabel(status: ReportJob['status']): string {
  const labels: Record<ReportJob['status'], string> = {
    pending: 'Pending',
    processing: 'Processing',
    completed: 'Completed',
    failed: 'Failed',
    cancelled: 'Cancelled',
  };
  return labels[status];
}

export function isJobComplete(job: ReportJob): boolean {
  return job.status === 'completed';
}

export function isJobInProgress(job: ReportJob): boolean {
  return job.status === 'pending' || job.status === 'processing';
}

export function createDefaultSection(): ReportSection {
  return {
    id: crypto.randomUUID(),
    title: 'New Section',
    elements: [],
    columns: 1,
    padding: 10,
  };
}

export function createDefaultBranding(): BrandingConfig {
  return {
    primary_color: '#1f2937',
    secondary_color: '#3b82f6',
    accent_color: '#10b981',
    font_family: 'Inter, sans-serif',
  };
}

export function createDefaultPageConfig(): PageConfig {
  return {
    size: 'a4',
    orientation: 'portrait',
    margins: { top: 20, bottom: 20, left: 15, right: 15 },
    header_footer: {
      show_header: true,
      show_footer: true,
      include_page_numbers: true,
      include_date: true,
    },
  };
}

File: backend/app/schemas/reports.py
"""
Report Builder Schemas

Pydantic schemas for report generation and export.
"""

from typing import Any, Optional, Union
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class ReportFormat(str, Enum):
    """Report export formats"""
    PDF = "pdf"
    HTML = "html"
    EXCEL = "excel"
    POWERPOINT = "pptx"
    IMAGE = "image"


class PageSize(str, Enum):
    """Page sizes for PDF"""
    A4 = "a4"
    LETTER = "letter"
    LEGAL = "legal"
    A3 = "a3"
    TABLOID = "tabloid"


class PageOrientation(str, Enum):
    """Page orientation"""
    PORTRAIT = "portrait"
    LANDSCAPE = "landscape"


class ElementType(str, Enum):
    """Report element types"""
    CHART = "chart"
    TABLE = "table"
    TEXT = "text"
    IMAGE = "image"
    KPI = "kpi"
    HEADING = "heading"
    DIVIDER = "divider"
    SPACER = "spacer"
    PAGE_BREAK = "page_break"
    TOC = "table_of_contents"


class TextAlignment(str, Enum):
    """Text alignment"""
    LEFT = "left"
    CENTER = "center"
    RIGHT = "right"
    JUSTIFY = "justify"


# Branding Configuration

class BrandingConfig(BaseModel):
    """Report branding configuration"""
    logo_url: Optional[str] = None
    logo_width: int = 150
    logo_position: str = "left"  # left, center, right
    primary_color: str = "#1f2937"
    secondary_color: str = "#3b82f6"
    accent_color: str = "#10b981"
    font_family: str = "Inter, sans-serif"
    header_font_family: Optional[str] = None
    company_name: Optional[str] = None
    footer_text: Optional[str] = None


# Page Configuration

class PageMargins(BaseModel):
    """Page margins in mm"""
    top: int = 20
    bottom: int = 20
    left: int = 15
    right: int = 15


class HeaderFooterConfig(BaseModel):
    """Header/footer configuration"""
    show_header: bool = True
    show_footer: bool = True
    header_height: int = 30
    footer_height: int = 20
    include_logo: bool = True
    include_title: bool = True
    include_date: bool = True
    include_page_numbers: bool = True
    custom_header_html: Optional[str] = None
    custom_footer_html: Optional[str] = None


class PageConfig(BaseModel):
    """Page configuration"""
    size: PageSize = PageSize.A4
    orientation: PageOrientation = PageOrientation.PORTRAIT
    margins: PageMargins = Field(default_factory=PageMargins)
    header_footer: HeaderFooterConfig = Field(default_factory=HeaderFooterConfig)


# Report Elements

class TextStyle(BaseModel):
    """Text styling"""
    font_size: int = 12
    font_weight: str = "normal"  # normal, bold
    font_style: str = "normal"  # normal, italic
    color: str = "#1f2937"
    alignment: TextAlignment = TextAlignment.LEFT
    line_height: float = 1.5


class ChartElement(BaseModel):
    """Chart element configuration"""
    element_type: str = "chart"
    chart_id: str
    title: Optional[str] = None
    show_title: bool = True
    width: str = "100%"
    height: int = 400
    filters: Optional[dict[str, Any]] = None


class TableElement(BaseModel):
    """Table element configuration"""
    element_type: str = "table"
    data_source: str  # query_id, dataset_id, or inline data
    query: Optional[str] = None
    columns: Optional[list[str]] = None
    title: Optional[str] = None
    show_title: bool = True
    max_rows: int = 100
    show_row_numbers: bool = False
    striped: bool = True
    bordered: bool = True
    compact: bool = False


class TextElement(BaseModel):
    """Text element configuration"""
    element_type: str = "text"
    content: str
    style: TextStyle = Field(default_factory=TextStyle)
    markdown: bool = False


class HeadingElement(BaseModel):
    """Heading element configuration"""
    element_type: str = "heading"
    text: str
    level: int = 1  # 1-6
    style: Optional[TextStyle] = None
    include_in_toc: bool = True


class ImageElement(BaseModel):
    """Image element configuration"""
    element_type: str = "image"
    url: str
    alt_text: str = ""
    width: Optional[str] = None
    height: Optional[str] = None
    alignment: TextAlignment = TextAlignment.CENTER


class KPIElement(BaseModel):
    """KPI card element"""
    element_type: str = "kpi"
    kpi_id: Optional[str] = None
    label: str
    value: Optional[Union[str, float]] = None
    query: Optional[str] = None  # Query to fetch value
    format: Optional[str] = None
    comparison_value: Optional[float] = None
    comparison_label: Optional[str] = None
    trend_direction: Optional[str] = None  # up, down, neutral
    color: Optional[str] = None


class DividerElement(BaseModel):
    """Divider element"""
    element_type: str = "divider"
    style: str = "solid"  # solid, dashed, dotted
    color: str = "#e5e7eb"
    thickness: int = 1
    margin_top: int = 10
    margin_bottom: int = 10


class SpacerElement(BaseModel):
    """Spacer element"""
    element_type: str = "spacer"
    height: int = 20


class PageBreakElement(BaseModel):
    """Page break element"""
    element_type: str = "page_break"


class TOCElement(BaseModel):
    """Table of contents element"""
    element_type: str = "table_of_contents"
    title: str = "Table of Contents"
    max_depth: int = 3
    show_page_numbers: bool = True


ReportElement = Union[
    ChartElement,
    TableElement,
    TextElement,
    HeadingElement,
    ImageElement,
    KPIElement,
    DividerElement,
    SpacerElement,
    PageBreakElement,
    TOCElement,
]


# Report Section

class ReportSection(BaseModel):
    """Report section"""
    id: str
    title: Optional[str] = None
    elements: list[ReportElement]
    columns: int = 1  # 1, 2, or 3 column layout
    background_color: Optional[str] = None
    padding: int = 10


# Report Template

class ReportTemplateBase(BaseModel):
    """Base report template schema"""
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None


class ReportTemplateCreate(ReportTemplateBase):
    """Schema for creating a report template"""
    page_config: PageConfig = Field(default_factory=PageConfig)
    branding: BrandingConfig = Field(default_factory=BrandingConfig)
    sections: list[ReportSection]
    workspace_id: Optional[str] = None
    is_public: bool = False
    tags: list[str] = Field(default_factory=list)


class ReportTemplateUpdate(BaseModel):
    """Schema for updating a report template"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    page_config: Optional[PageConfig] = None
    branding: Optional[BrandingConfig] = None
    sections: Optional[list[ReportSection]] = None
    is_public: Optional[bool] = None
    tags: Optional[list[str]] = None


class ReportTemplateResponse(ReportTemplateBase):
    """Schema for report template response"""
    id: str
    page_config: PageConfig
    branding: BrandingConfig
    sections: list[ReportSection]
    workspace_id: Optional[str] = None
    created_by: str
    created_at: datetime
    updated_at: datetime
    is_public: bool
    usage_count: int = 0
    tags: list[str] = Field(default_factory=list)

    class Config:
        from_attributes = True


# Report Generation

class ReportGenerateRequest(BaseModel):
    """Request to generate a report"""
    template_id: Optional[str] = None
    dashboard_id: Optional[str] = None  # Generate from dashboard
    title: str = "Report"
    subtitle: Optional[str] = None
    format: ReportFormat = ReportFormat.PDF
    page_config: Optional[PageConfig] = None
    branding: Optional[BrandingConfig] = None
    sections: Optional[list[ReportSection]] = None
    filters: Optional[dict[str, Any]] = None  # Apply filters to data
    parameters: Optional[dict[str, Any]] = None  # Template parameters
    include_cover_page: bool = True
    include_toc: bool = False
    generated_date: Optional[datetime] = None


class ReportGenerateResponse(BaseModel):
    """Response for report generation"""
    job_id: str
    status: str  # pending, processing, completed, failed
    format: ReportFormat
    message: Optional[str] = None
    download_url: Optional[str] = None
    expires_at: Optional[datetime] = None


class ReportJob(BaseModel):
    """Report generation job"""
    id: str
    template_id: Optional[str] = None
    dashboard_id: Optional[str] = None
    title: str
    format: ReportFormat
    status: str
    progress: int = 0
    created_by: str
    created_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None
    download_url: Optional[str] = None
    file_size: Optional[int] = None
    expires_at: Optional[datetime] = None


# Scheduled Reports

class ScheduledReportConfig(BaseModel):
    """Scheduled report configuration"""
    template_id: str
    title: str
    format: ReportFormat = ReportFormat.PDF
    filters: Optional[dict[str, Any]] = None
    parameters: Optional[dict[str, Any]] = None


class ScheduledReport(BaseModel):
    """Scheduled report"""
    id: str
    name: str
    config: ScheduledReportConfig
    schedule: str  # Cron expression
    recipients: list[str]
    enabled: bool = True
    workspace_id: Optional[str] = None
    created_by: str
    created_at: datetime
    last_generated_at: Optional[datetime] = None
    next_generation_at: Optional[datetime] = None


# Export Options

class PDFExportOptions(BaseModel):
    """PDF-specific export options"""
    page_size: PageSize = PageSize.A4
    orientation: PageOrientation = PageOrientation.PORTRAIT
    margins: PageMargins = Field(default_factory=PageMargins)
    print_background: bool = True
    display_header_footer: bool = True
    scale: float = 1.0


class ExcelExportOptions(BaseModel):
    """Excel-specific export options"""
    include_charts_as_images: bool = True
    include_formulas: bool = False
    freeze_headers: bool = True
    auto_filter: bool = True
    sheet_per_section: bool = False


class ImageExportOptions(BaseModel):
    """Image export options"""
    format: str = "png"  # png, jpeg, webp
    width: int = 1200
    height: Optional[int] = None  # Auto-calculate
    scale: float = 2.0  # For retina
    quality: int = 90  # For jpeg


# Dashboard Export

class DashboardExportRequest(BaseModel):
    """Request to export a dashboard"""
    dashboard_id: str
    format: ReportFormat = ReportFormat.PDF
    title: Optional[str] = None
    include_filters: bool = True
    filter_state: Optional[dict[str, Any]] = None
    pdf_options: Optional[PDFExportOptions] = None
    excel_options: Optional[ExcelExportOptions] = None
    image_options: Optional[ImageExportOptions] = None


class ChartExportRequest(BaseModel):
    """Request to export a single chart"""
    chart_id: str
    format: str = "png"  # png, jpeg, svg, pdf
    width: int = 800
    height: int = 600
    scale: float = 2.0
    include_title: bool = True
    include_legend: bool = True
    background_color: str = "white"


# Template Library

class TemplateCategory(BaseModel):
    """Template category"""
    id: str
    name: str
    description: Optional[str] = None
    icon: Optional[str] = None
    template_count: int = 0


class TemplatePreview(BaseModel):
    """Template preview info"""
    id: str
    name: str
    description: Optional[str] = None
    thumbnail_url: Optional[str] = None
    category: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    usage_count: int = 0
    is_public: bool = False


# Constants

BUILT_IN_TEMPLATES = [
    {
        "id": "executive-summary",
        "name": "Executive Summary",
        "description": "High-level overview with KPIs and key charts",
        "category": "business",
        "tags": ["executive", "summary", "kpi"],
    },
    {
        "id": "financial-report",
        "name": "Financial Report",
        "description": "Detailed financial analysis with tables and trends",
        "category": "finance",
        "tags": ["finance", "accounting", "budget"],
    },
    {
        "id": "sales-performance",
        "name": "Sales Performance",
        "description": "Sales metrics, pipeline, and team performance",
        "category": "sales",
        "tags": ["sales", "revenue", "pipeline"],
    },
    {
        "id": "marketing-analytics",
        "name": "Marketing Analytics",
        "description": "Campaign performance and marketing metrics",
        "category": "marketing",
        "tags": ["marketing", "campaigns", "roi"],
    },
    {
        "id": "operational-dashboard",
        "name": "Operational Dashboard",
        "description": "Real-time operational metrics and KPIs",
        "category": "operations",
        "tags": ["operations", "metrics", "monitoring"],
    },
]

FORMAT_EXTENSIONS = {
    ReportFormat.PDF: ".pdf",
    ReportFormat.HTML: ".html",
    ReportFormat.EXCEL: ".xlsx",
    ReportFormat.POWERPOINT: ".pptx",
    ReportFormat.IMAGE: ".png",
}

FORMAT_MIME_TYPES = {
    ReportFormat.PDF: "application/pdf",
    ReportFormat.HTML: "text/html",
    ReportFormat.EXCEL: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    ReportFormat.POWERPOINT: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    ReportFormat.IMAGE: "image/png",
}
